lib/src/js/value/color.dart

```diff
@@ -2,66 +2,351 @@
-        if (options.red case var red?) "red": red,
-        if (options.green case var green?) "green": green,
-        if (options.blue case var blue?) "blue": blue,
-        if (options.alpha case var alpha?) "alpha": alpha
-      });
-    } else {
-      return self.changeAlpha(options.alpha ?? self.alpha);
+  jsClass.defineMethods({
+    'equals': (SassColor self, Object other) => self == other,
+    'hashCode': (SassColor self) => self.hashCode,
+    'toSpace': (SassColor self, String space) => _toSpace(self, space),
+    'isInGamut': (SassColor self, [String? space]) =>
+        _toSpace(self, space).isInGamut,
+    'toGamut': (SassColor self, [String? space]) =>
+        _toSpace(self, space).toGamut(),
+    'channel': (SassColor self, String channel, [_ChannelOptions? options]) =>
+        _toSpace(self, options?.space).channel(channel),
+    'isChannelMissing': (SassColor self, String channel) =>
+        self.isChannelMissing(channel),
+    'isChannelPowerless': (SassColor self, String channel,
+            [_ChannelOptions? options]) =>
+        _toSpace(self, options?.space).isChannelPowerless(channel),
+    'change': (SassColor self, _ConstructionOptions options) {
+      var spaceSetExplicitly = options.space != null;
+      var space =
+          spaceSetExplicitly ? ColorSpace.fromName(options.space!) : self.space;
+
+      if (self.isLegacy && !spaceSetExplicitly) {
+        if (hasProperty(options, 'whiteness') ||
+            hasProperty(options, 'blackness')) {
+          space = ColorSpace.hwb;
+        } else if (hasProperty(options, 'hue') ||
+            hasProperty(options, 'saturation') ||
+            hasProperty(options, 'lightness')) {
+          space = ColorSpace.hsl;
+        } else if (hasProperty(options, 'red')) {
+          space = ColorSpace.rgb;
+        }
+        if (space != self.space) {
+          warnForDeprecationFromApi(
+              "Changing a channel not in this color's space without explicitly specifying "
+              "the `space` option is deprecated.",
+              Deprecation.color4Api);
+        }
+      }
+
+      for (final key in objectKeys(options)) {
+        if (['alpha', 'space'].contains(key)) continue;
+        if (!space.channels.any((channel) => channel.name == key)) {
+          jsThrow(JsError("`$key` is not a valid channel in `$space`."));
+        }
+      }
+
+      var color = self.toSpace(space);
+
+      SassColor changedColor;
+
+      changedValue(String channel) {
+        return _changeComponentValue(color, channel, options);
+      }
+
+      switch (space) {
+        case ColorSpace.hsl when spaceSetExplicitly:
+          changedColor = SassColor.hsl(
+              changedValue('hue'),
+              changedValue('saturation'),
+              changedValue('lightness'),
+              changedValue('alpha'));
+          break;
+
+        case ColorSpace.hsl:
+          if ((hasProperty(options, 'hue') && options.hue == null) ||
```

Comment:
This will still return true if the user explicitly sets the property to `undefined` rather than just omitting it. If you want to check if it's specifically the value `null`, you'll need to do `identical(options.hue, null)`.

There are three distinct states a property can be in that all look like `options.hue == null` to Dart:

1. Absent (`{}`)
2. Explicitly undefined (`{hue: undefined}`)
3. Explicitly null (`{hue: null}`)

We generally want the first two states to work the same way, and for cases where we support missing channels we want the third to work differently. You can check for each of them as follows, respectively:

1. `!hasProperty(options, 'hue')`
2. `hasProperty(options, 'hue') && isUndefined(options.hue)`
3. `identical(options.hue, null)`

...but since you usually want to have the same behavior for the first two cases, `isUndefined(options.hue)` or `identical(options.hue, null)` are usually the checks you want.

Reply:
In running some tests, I found that `identical(options.hue, null)` is true if `options.hue` is explicitly undefined. What I was able to determine for each of the checks was this-

1. Absent- `!hasProperty(options, 'hue')`
2. Explicitly undefined- `hasProperty(options, 'hue') && isUndefined(options.hue)`,
3. Explicitly null - `!isUndefined(options.hue) && identical(options.hue, null)`

I also found that `identical(options.hue, null)` was identical to `options.hue == null` in all cases I tested. Does this track with your expectations?

Reply:
Add a JS util `isExplicitlyNull` function for this.

Reply:
Added `isNull` (to match `isUndefined`) in 632df8cd4908441bc0732a3ae710a7b91fb00f91.